src/dune_lang/slang.ml

```diff
@@ -58,12 +58,37 @@ let decode =
             Or_absorb_undefined_var x )
         ]
     in
-    located decode_form
-    >>| (fun (loc, x) -> Form (loc, x))
-    <|> (let+ x = String_with_vars.decode in
-         Literal x)
-    <|> let+ loc, x = located (Blang.Ast.decode decode) in
-        Form (loc, Blang x))
+    try_
+      (let+ loc, x = located decode_form in
+       Form (loc, x))
+      (fun slang_exn ->
+        let decode_literal =
+          let+ x = String_with_vars.decode in
+          Literal x
+        in
+        let decode_blang =
+          let* blang_start = loc in
+          let decode_blang_string =
+            (* The decoders for the blang and slang DSLs are mutually recursive
+               since blang expressions can appear as the conditions in slang
+               expressions and slang expressions can be compared in blang
+               expressions, and also produce blang literals. When encountering
+               a form which is not valid syntax for slang nor blang each
+               decoder will attempt to invoke the other leading to infinite
+               recursion. To prevent this, record the decoder location before
+               attempting to parse the blang expression. If the blang decoder
+               invokes the slang decoder at the same location we know that it's
+               not a valid blang expression. Since we only invoke the blang
+               decoder when the slang decoder has already failed to interpret
+               the form, the form is invalid so raise the original exception
+               from when the slang decoder failed to interpret it. *)
+            let* string_start = loc in
+            if Loc.equal string_start blang_start then raise slang_exn else decode
+          in
+          let+ loc, x = located (Blang.Ast.decode decode_blang_string) in
```

Comment:
It feels weird to me to solve this with a comparison of Loc. Maybe this could be solved by a parameter `recurse_blang`/`recurse_slang` to detect a Blang->Slang->Blang or Slang->Blang->Slang case?

Reply:
Good point. I've added a `Blang.Ast.decode_without_bare_literals` decoder for blang which doesn't parse boolean literals to break cycles like this without resorting to comparing locations.

Reply:
That seems like a much less "magical" way, much better! However it seems it triggers new test failures.